File: apple/Peacock/Protocol/Payloads.swift
import Foundation

// MARK: - Announce

struct AnnouncePayload: Sendable {
    let deviceName: String
    let platform: String
    let tcpPort: UInt16
    let features: UInt32
    let restrictedPeers: [PeerInfo]

    func encode() -> Data {
        let enc = BincodeEncoder()
        enc.encodeString(deviceName)
        enc.encodeString(platform)
        enc.encodeU16(tcpPort)
        enc.encodeU32(features)
        enc.encodeU64(UInt64(restrictedPeers.count))
        for peer in restrictedPeers {
            peer.encode(to: enc)
        }
        return enc.data
    }

    static func decode(from data: Data) throws -> AnnouncePayload {
        let dec = BincodeDecoder(data: data)
        let deviceName = try dec.decodeString()
        let platform = try dec.decodeString()
        let tcpPort = try dec.decodeU16()
        let features = try dec.decodeU32()
        let peerCount = try dec.decodeU64()
        var peers: [PeerInfo] = []
        for _ in 0..<peerCount {
            peers.append(try PeerInfo.decode(from: dec))
        }
        return AnnouncePayload(deviceName: deviceName, platform: platform,
                               tcpPort: tcpPort, features: features,
                               restrictedPeers: peers)
    }
}

struct PeerInfo: Sendable {
    let deviceId: String
    let deviceName: String
    let ipAddr: String
    let tcpPort: UInt16
    let platform: String

    func encode(to enc: BincodeEncoder) {
        enc.encodeString(deviceId)
        enc.encodeString(deviceName)
        enc.encodeString(ipAddr)
        enc.encodeU16(tcpPort)
        enc.encodeString(platform)
    }

    static func decode(from dec: BincodeDecoder) throws -> PeerInfo {
        PeerInfo(
            deviceId: try dec.decodeString(),
            deviceName: try dec.decodeString(),
            ipAddr: try dec.decodeString(),
            tcpPort: try dec.decodeU16(),
            platform: try dec.decodeString()
        )
    }
}

// MARK: - Text

struct TextPayload: Sendable {
    let messageId: String
    let text: String
    let timestamp: UInt64

    func encode() -> Data {
        let enc = BincodeEncoder()
        enc.encodeString(messageId)
        enc.encodeString(text)
        enc.encodeU64(timestamp)
        return enc.data
    }

    static func decode(from data: Data) throws -> TextPayload {
        let dec = BincodeDecoder(data: data)
        return TextPayload(
            messageId: try dec.decodeString(),
            text: try dec.decodeString(),
            timestamp: try dec.decodeU64()
        )
    }
}

// MARK: - File Transfer

struct FileOfferPayload: Sendable {
    let transferId: String
    let fileName: String
    let fileSize: UInt64
    let isFolder: Bool
    let fileCount: UInt32

    func encode() -> Data {
        let enc = BincodeEncoder()
        enc.encodeString(transferId)
        enc.encodeString(fileName)
        enc.encodeU64(fileSize)
        enc.encodeBool(isFolder)
        enc.encodeU32(fileCount)
        return enc.data
    }

    static func decode(from data: Data) throws -> FileOfferPayload {
        let dec = BincodeDecoder(data: data)
        return FileOfferPayload(
            transferId: try dec.decodeString(),
            fileName: try dec.decodeString(),
            fileSize: try dec.decodeU64(),
            isFolder: try dec.decodeBool(),
            fileCount: try dec.decodeU32()
        )
    }
}

struct FileAcceptPayload: Sendable {
    let transferId: String
    let receiverPort: UInt16
    let resumeOffset: UInt64

    func encode() -> Data {
        let enc = BincodeEncoder()
        enc.encodeString(transferId)
        enc.encodeU16(receiverPort)
        enc.encodeU64(resumeOffset)
        return enc.data
    }

    static func decode(from data: Data) throws -> FileAcceptPayload {
        let dec = BincodeDecoder(data: data)
        return FileAcceptPayload(
            transferId: try dec.decodeString(),
            receiverPort: try dec.decodeU16(),
            resumeOffset: try dec.decodeU64()
        )
    }
}

struct FileRejectPayload: Sendable {
    let transferId: String

    func encode() -> Data {
        let enc = BincodeEncoder()
        enc.encodeString(transferId)
        return enc.data
    }

    static func decode(from data: Data) throws -> FileRejectPayload {
        let dec = BincodeDecoder(data: data)
        return FileRejectPayload(transferId: try dec.decodeString())
    }
}

// MARK: - Snippet

struct SnippetSharePayload: Sendable {
    let title: String
    let content: String
    let tag: String
    let note: String

    func encode() -> Data {
        let enc = BincodeEncoder()
        enc.encodeString(title)
        enc.encodeString(content)
        enc.encodeString(tag)
        enc.encodeString(note)
        return enc.data
    }

    static func decode(from data: Data) throws -> SnippetSharePayload {
        let dec = BincodeDecoder(data: data)
        return SnippetSharePayload(
            title: try dec.decodeString(),
            content: try dec.decodeString(),
            tag: try dec.decodeString(),
            note: try dec.decodeString()
        )
    }
}

File: android/app/src/main/java/com/peacock/app/protocol/Payloads.kt
package com.peacock.app.protocol

/**
 * All payload types — field order must match Rust's bincode serialization exactly.
 */

data class PeerInfo(
    val deviceId: String,
    val deviceName: String,
    val ipAddr: String,
    val tcpPort: UShort,
    val platform: String
) {
    fun encode(): ByteArray = BincodeEncoder()
        .encodeString(deviceId)
        .encodeString(deviceName)
        .encodeString(ipAddr)
        .encodeU16(tcpPort)
        .encodeString(platform)
        .toByteArray()

    companion object {
        fun decode(decoder: BincodeDecoder): PeerInfo = PeerInfo(
            deviceId = decoder.decodeString(),
            deviceName = decoder.decodeString(),
            ipAddr = decoder.decodeString(),
            tcpPort = decoder.decodeU16(),
            platform = decoder.decodeString()
        )
    }
}

data class AnnouncePayload(
    val deviceName: String,
    val platform: String,
    val tcpPort: UShort,
    val features: UInt,
    val restrictedPeers: List<PeerInfo> = emptyList()
) {
    fun encode(): ByteArray {
        val enc = BincodeEncoder()
            .encodeString(deviceName)
            .encodeString(platform)
            .encodeU16(tcpPort)
            .encodeU32(features)
            .encodeU64(restrictedPeers.size.toULong())
        for (peer in restrictedPeers) {
            enc.encodeBytes(peer.encode())
        }
        return enc.toByteArray()
    }

    companion object {
        fun decode(data: ByteArray): AnnouncePayload {
            val d = BincodeDecoder(data)
            val deviceName = d.decodeString()
            val platform = d.decodeString()
            val tcpPort = d.decodeU16()
            val features = d.decodeU32()
            val peerCount = d.decodeU64().toInt()
            val peers = (0 until peerCount).map { PeerInfo.decode(d) }
            return AnnouncePayload(deviceName, platform, tcpPort, features, peers)
        }
    }
}

data class TextPayload(
    val messageId: String,
    val text: String,
    val timestamp: ULong
) {
    fun encode(): ByteArray = BincodeEncoder()
        .encodeString(messageId)
        .encodeString(text)
        .encodeU64(timestamp)
        .toByteArray()

    companion object {
        fun decode(data: ByteArray): TextPayload {
            val d = BincodeDecoder(data)
            return TextPayload(d.decodeString(), d.decodeString(), d.decodeU64())
        }
    }
}

data class FileOfferPayload(
    val transferId: String,
    val fileName: String,
    val fileSize: ULong,
    val isFolder: Boolean,
    val fileCount: UInt
) {
    fun encode(): ByteArray = BincodeEncoder()
        .encodeString(transferId)
        .encodeString(fileName)
        .encodeU64(fileSize)
        .encodeBool(isFolder)
        .encodeU32(fileCount)
        .toByteArray()

    companion object {
        fun decode(data: ByteArray): FileOfferPayload {
            val d = BincodeDecoder(data)
            return FileOfferPayload(
                d.decodeString(), d.decodeString(), d.decodeU64(),
                d.decodeBool(), d.decodeU32()
            )
        }
    }
}

data class FileAcceptPayload(
    val transferId: String,
    val receiverPort: UShort,
    val resumeOffset: ULong
) {
    fun encode(): ByteArray = BincodeEncoder()
        .encodeString(transferId)
        .encodeU16(receiverPort)
        .encodeU64(resumeOffset)
        .toByteArray()

    companion object {
        fun decode(data: ByteArray): FileAcceptPayload {
            val d = BincodeDecoder(data)
            return FileAcceptPayload(d.decodeString(), d.decodeU16(), d.decodeU64())
        }
    }
}

data class FileRejectPayload(
    val transferId: String
) {
    fun encode(): ByteArray = BincodeEncoder()
        .encodeString(transferId)
        .toByteArray()

    companion object {
        fun decode(data: ByteArray): FileRejectPayload {
            val d = BincodeDecoder(data)
            return FileRejectPayload(d.decodeString())
        }
    }
}

data class SnippetSharePayload(
    val title: String,
    val content: String,
    val tag: String,
    val note: String
) {
    fun encode(): ByteArray = BincodeEncoder()
        .encodeString(title)
        .encodeString(content)
        .encodeString(tag)
        .encodeString(note)
        .toByteArray()

    companion object {
        fun decode(data: ByteArray): SnippetSharePayload {
            val d = BincodeDecoder(data)
            return SnippetSharePayload(
                d.decodeString(), d.decodeString(),
                d.decodeString(), d.decodeString()
            )
        }
    }
}
